Treat methods with empty bodies in Protocols as abstract

thomkeh opened the pull request:
### Description

closes #8005
closes #8926

Based on the _quick fix_ idea [from this comment](https://github.com/python/mypy/issues/8005#issuecomment-558341465) by @ilevkivskyi . 

For the problem description see #8005. The underlying problem is arguably #2350, but that seems not so easy to fix (there was an attempt in #8111 but it was abandoned). For this PR, I borrowed a lot from #8111 but kept the scope much less ambitious (for example, #8111 also warns on `super()` calls to empty functions, which this PR doesn't).  However, as [pointed out here](https://github.com/python/mypy/issues/2350#issuecomment-256701609), empty bodies are actually a valid implementation if the function may return `None`, so I added that as an additional check. But if that is deemed too confusing, I'm happy to remove that check.

So, in summary, what this PR does: Methods in Protocols are considered abstract if they
1. have an empty function body
2. have a return type that is not compatible with `None` and
3. are not in a stub file.

## Test Plan

Added tests for all the new behavior.

nickgaya commented:
Issues #8005 and #8926 are both based on an incorrect user belief that an empty method in a Protocol is abstract.  However this does not match Python's runtime behavior.

Instead of treating empty methods as abstract, MyPy should flag the method as missing a return statement, and optionally suggest adding the `@abstractmethod` annotation.

For example, consider this class definition:

    class P(protocol):
        def foo() -> int: ...

MyPy should raise an error like this:

    # error: Missing return statement
    # note: Did you mean to use @abstractmethod?

sobolevn commented on mypy/semanal.py:
```diff
@@ -803,6 +813,21 @@ def analyze_overloaded_func_def(self, defn: OverloadedFuncDef) -> None:
+        Currently, this checks for a trivial body in protocols classes,
+        where it makes the method implicitly abstract.
+        """
+        assert defn.impl is not None
```
nit: I think that it is better to use `if defn.impl is None: return` here.

And remove this check: https://github.com/python/mypy/pull/12118/files#diff-f782aad4853edceb67a88efe43d2869579d075a77b8e8dd2add91e818a9fa164R816

This way we won't have accidental `AssertionError` if someone uses this method incorrectly.

thomkeh replied:
That's of course fine by me, but note that I just moved this function and didn't modify it in any way.

thomkeh replied:
Sorry, I was confused about which function you were talking about. This is not the function I moved. I fixed it.

sobolevn reviewed (commented):
First of all, thanks for working on this!

But, sorry, I fail to understand this feature.

Questions I have:
1. I was always sure that we use protocols to define some object's *shape*, not ever use them to define *behavior*. So, I was always thinking that all protocol methods are abstract. (Moreover, I think that they should be forced to be empty, but that's a completely other discussion). Take a look at `interface` in TypeScript.
2. Why a method with `-> None` is not treated as `abstract`?
3. I think that in current state this will cause more problems that solve. For example, it would quite hard to teach which methods are abstract and which are not. I would appreciate a plan on how we can do that

a-reich commented:
> First of all, thanks for working on this!
> 
> But, sorry, I fail to understand this feature.
> 
> Questions I have:
> 
> 1. I was always sure that we use protocols to define some object's _shape_, not ever use them to define _behavior_. So, I was always thinking that all protocol methods are abstract. (Moreover, I think that they should be forced to be empty, but that's a completely different discussion).

That may be how most people think of protocols, but PEP 544 chose to state that they can provide non-abstract “default implementations” and classes can rely on those. Which I think has now led to this confusion about what it means to subclass a proto and how to check it.

thomkeh commented:
So, my goal was to make mypy raise an error for this case:

```python
from typing import Protocol

class Sized(Protocol):
    def __len__(self) -> int: ...

class A(Sized): ...
a = A()
assert isinstance(len(a), int)  # runtime error but no mypy error
```

The problem here is the interplay of explicit protocol-implementation (i.e., subclassing of `Sized`) and the possibility of default implementations. Without either of those things, there would be no problem. But, [the PEP](https://www.python.org/dev/peps/pep-0544/) allows for both, so here we are.

One possible solution would indeed be to force protocol methods to either provide a correct default implementation (i.e., one that returns the correct value) or to be marked as abstract:

```python
from abc import abstractmethod
from typing import Protocol

class P(Protocol):
    def meth1(self) -> int:  # E: Missing return statement
        pass

    @abstractmethod
    def meth2(self) -> int:  # OK
        pass
```
But I feel that goes a bit against what protocols were meant to do? I feel like default implementations in general are kind of not a good fit for protocols; they're better for abstract base classes, but I don't know. The advantage of this approach is that it works in stub files as well.

The other solution is to treat protocol methods that don't return what they promise to return as implicitly abstract. (Or maybe "abstract" is not the right term. Because as @nickgaya points out, protocol methods with empty default implementations aren't really "abstract" in the sense of abstract base classes; the problem is just that they implicitly return `None` (because they don't have a `return` statement) and thus should maybe simply cause a type error.) The disadvantage of this approach is that it doesn't work in stub files.

thomkeh commented:
Hmm, I realize now that mypy in general doesn't complain about empty function bodies: #2350 (there was an attempt to fix it in  #8111 but it was abandoned), but I still think the problem is especially bad for protocols, so this PR should be helpful.

thomkeh commented:
I addressed the review feedback.

There is some new relevant discussion in #8926.

Would it be better to remove the check of the return type? This would then be more in line with what pyright does.

thomkeh commented:
Any interest in this? @ilevkivskyi

ilevkivskyi reviewed (commented):
I like this PR. Issues with empty bodies come up quite often. About special-casing `None`: I don't think this is a good idea. In general, few people know that protocols (as well as ordinary ABCs) can have default implementations. And even among those who know and use them, few will have empty body as an _intended_ default implementation.

Also semantic analysis is a wrong place for type-checking. The current implementation already misses couple corner cases, like type variables that can be `None`, and maybe there are more.

I think it will be better to give a note that an explicit `return None` must be used if it is intentional. (IIRC we give instantiation error at type-checking stage, where you can simply use `is_subtype()`)

thomkeh commented:
> About special-casing `None`: I don't think this is a good idea. In general, few people know that protocols (as well as ordinary ABCs) can have default implementations. And even among those who know and use them, few will have empty body as an _intended_ default implementation.

Yeah, that makes sense – I'll remove that special-case-ing.

> Also semantic analysis is a wrong place for type-checking. The current implementation already misses couple corner cases, like type variables that can be `None`, and maybe there are more.

I see. I will try to find the right place.

> I think it will be better to give a note that an explicit `return None` must be used if it is intentional. (IIRC we give instantiation error at type-checking stage, where you can simply use `is_subtype()`)

That sounds good. I will try to figure out how to do that.

thomkeh commented:
@ilevkivskyi I implemented your suggestions (hopefully I understood them correctly), but there is a problem with `strict_optional` that I don't know how to solve. When `strict_optional` is `False`, then `is_subtype()` is basically useless when the first argument is `NoneType`. This means, it's not possible to show the helpful note in that case. I guess I could modify `is_subtype()` so that you can turn off the `strict_optional` behavior with an additional argument.

The other problem is that the `_check_non_ret_from_prot()` function is quite complicated and recomputes things that are known at other places in the code, but I think it's probably not worth caching this information from those places, because it's only needed for an error message that should be quite rare.

I'm also confused why I had to add `if not typ.is_protocol:` in `semanal_classprop.py` but without that, the stubgen tests were failing. However, it should have been previously already possible to create an abstract protocol by explicitly using `@abc.abstractmethod`.

I didn't use an `Enum` for `NOT_ABSTRACT`/`IS_ABSTRACT`/`IMPLICITLY_ABSTRACT` because to my knowledge, enums are not optimized well by mypyc, but I can change that if you want.

ilevkivskyi commented on mypy/checker.py:
```diff
@@ -618,7 +617,7 @@ def _visit_overloaded_func_def(self, defn: OverloadedFuncDef) -> None:
             assert isinstance(fdef, Decorator)
             self.check_func_item(fdef.func, name=fdef.func.name)
-            if fdef.func.is_abstract:
+            if fdef.func.abstract_status:
```
Please don't rely on specific constants being falsey. Use explicit `==` instead. Otherwise if someone will change numeric value of the constant this code will subtly break.

ilevkivskyi commented on mypy/strconv.py:
```diff
@@ -133,7 +133,7 @@ def visit_func_def(self, o: "mypy.nodes.FuncDef") -> str:
         if len(arg_kinds & {mypy.nodes.ARG_NAMED, mypy.nodes.ARG_NAMED_OPT}) > 0:
             a.insert(1, f"MaxPos({o.max_pos})")
-        if o.is_abstract:
+        if o.abstract_status:
```
Please don't rely on specific constants being falsey. Use explicit == instead. Also please check all other places.

ilevkivskyi commented on mypy/semanal.py:
```diff
@@ -1052,12 +1083,13 @@ def handle_missing_overload_implementation(self, defn: OverloadedFuncDef) -> Non
             if self.type and self.type.is_protocol and not self.is_func_scope():
-                # An overloaded protocol method doesn't need an implementation.
+                # An overloaded protocol method doesn't need an implementation,
+                # but if it doesn't have one, then it is considered implicitly abstract.
```
I think an overload without implementation at all should be explicitly abstract, so you should use `IS_ABSTRACT` below.

ilevkivskyi commented on mypy/checkexpr.py:
```diff
@@ -1335,6 +1345,46 @@ def check_callable_call(
+                    return is_subtype(NoneType(), node.type.ret_type)
+                else:
+                    # No type annotations are available
+                    return True
```
I think we should return False here, don't add the note if we are not sure what is the intended return type. You can simply delete the `else` branch so we fall through to last `return False`.

ilevkivskyi commented on mypy/checkexpr.py:
```diff
@@ -1335,6 +1345,46 @@ def check_callable_call(
+                continue
+            node = symnode.node
+            if isinstance(node, OverloadedFuncDef):
+                if node.impl is not None:
```
As I said above, overloads without implementation are explicitly abstract. So this whole block can be turned into `node = node.impl`

ilevkivskyi commented on mypy/messages.py:
```diff
@@ -1330,6 +1330,15 @@ def cannot_instantiate_abstract_class(
             context,
             code=codes.ABSTRACT,
         )
+        for a, is_none_ret_and_prot in abstract_attributes.items():
```
Use a simpler name for this, like `can_return_none`.

ilevkivskyi commented on mypy/checkexpr.py:
```diff
@@ -1335,6 +1345,46 @@ def check_callable_call(
             callee = callee.copy_modified(ret_type=new_ret_type)
         return callee.ret_type, callee
 
+    def _check_non_ret_from_prot(self, type: TypeInfo, attr_name: str) -> bool:
```
Give this a better name like `can_return_none_in_protocol_base`.

ilevkivskyi commented on mypy/checkexpr.py:
```diff
@@ -1236,24 +1238,32 @@ def check_callable_call(
+                    continue
+                is_none_ret_from_prot[attr_name] = self._check_non_ret_from_prot(type, attr_name)
+            self.msg.cannot_instantiate_abstract_class(
+                callee.type_object().name, is_none_ret_from_prot, context
```
Just call this `abstract_attributes` (as the function parameter).

ilevkivskyi commented on mypy/nodes.py:
```diff
@@ -758,7 +758,12 @@ def is_dynamic(self) -> bool:
+# Abstract status of a function
+NOT_ABSTRACT: Final = 0
+IS_ABSTRACT: Final = 1
+IMPLICITLY_ABSTRACT: Final = 2
```
Add short comment above explaining the meaning of this value.

ilevkivskyi commented on mypy/nodes.py:
```diff
@@ -2674,7 +2681,7 @@ class is generic then it will be a type constructor of higher kind.
     is_protocol: bool  # Is this a protocol class?
     runtime_protocol: bool  # Does this protocol support isinstance checks?
-    abstract_attributes: List[str]
+    abstract_attributes: List[Tuple[str, int]]
```
Add a comment above explaining what is the meaning of this attribute (i.e. what is `int` there).

ilevkivskyi reviewed (commented):
Thanks! This looks better, I have some more detailed comments now.

ilevkivskyi commented:
> I'm also confused why I had to add `if not typ.is_protocol:` in `semanal_classprop.py` but without that, the stubgen tests were failing. However, it should have been previously already possible to create an abstract protocol by explicitly using `@abc.abstractmethod`.

Maybe stubgen automatically adds some imports if it sees an abstract class. In any case, I wouldn't touch `semanal_classprop` for this, fix the stubgen (or just update the test cases, if there are only few) instead.

thomkeh commented:
I think I addressed everything. The code looks cleaner now, I think.

The only potential problem I see now is the `strict_optional` issue.

ilevkivskyi reviewed (approved):
I wouldn't worry too much about non-strict optional. Few people use it now, and it is just a usability issue, not correctness.

I have one more suggestion, after you fix it I will merge when tests pass.

thomkeh commented on mypy/messages.py:
```diff
@@ -1330,6 +1330,15 @@ def cannot_instantiate_abstract_class(
             context,
             code=codes.ABSTRACT,
         )
+        for a, can_return_none_and_implicit in abstract_attributes.items():
```
Done. If you have suggestions on re-wording the message in any way, let me know.
